pkg/sbproxy/reconcile/reconcile_visibilities.go

```diff
@@ -0,0 +1,351 @@
+			&platform.ServiceVisibilityEntity{
+				Public:        true,
+				CatalogPlanID: smPlan.CatalogID,
+				Labels:        map[string]string{},
+			},
+		}
+	}
+
+	scopeLabelIndex := findOrgLabelIndex(visibility.Labels, scopeLabelKey)
+	if scopeLabelIndex == -1 {
+		return []*platform.ServiceVisibilityEntity{}
+	}
+
+	scopes := visibility.Labels[scopeLabelIndex].Value
+	result := make([]*platform.ServiceVisibilityEntity, 0, len(scopes))
+	for _, scope := range scopes {
+		result = append(result, &platform.ServiceVisibilityEntity{
+			Public:        false,
+			CatalogPlanID: smPlan.CatalogID,
+			Labels:        map[string]string{scopeLabelKey: scope},
+		})
+	}
+	return result
+}
+
+func findOrgLabelIndex(labels []*types.VisibilityLabel, labelKey string) int {
+	for i, label := range labels {
+		if label.Key == labelKey {
+			return i
+		}
+	}
+	return -1
+}
+
+func (r ReconciliationTask) reconcileServiceVisibilities(platformVis, smVis []*platform.ServiceVisibilityEntity) bool {
+	logger := log.C(r.ctx)
+	logger.Debug("ReconciliationTask reconsiling platform and SM visibilities...")
+
+	platformMap := r.convertVisListToMap(platformVis)
+	visibilitiesToCreate := make([]*platform.ServiceVisibilityEntity, 0)
+	for _, visibility := range smVis {
+		key := r.getVisibilityKey(visibility)
+		existingVis := platformMap[key]
+		delete(platformMap, key)
+		if existingVis == nil {
+			visibilitiesToCreate = append(visibilitiesToCreate, visibility)
+		}
+	}
+
+	logger.Debugf("ReconciliationTask %d visibilities will be removed from the platform", len(platformMap))
+	errorOccured := r.deleteVisibilities(platformMap)
+
+	logger.Debugf("ReconciliationTask %d visibilities will be created in the platform", len(visibilitiesToCreate))
+	errorOccured = r.createVisibilities(visibilitiesToCreate) || errorOccured
+
+	return errorOccured
+}
+
+// deleteVisibilities deletes visibilities from platform. Returns true if error has occurred
+func (r ReconciliationTask) deleteVisibilities(visibilities map[string]*platform.ServiceVisibilityEntity) bool {
+	var errorOccured int32 // will be used as atomic bool
+	var wg sync.WaitGroup
+
+	for _, visibility := range visibilities {
+		execAsync(&wg, visibility, &errorOccured, r.deleteVisibility)
+	}
+	return await(&wg, &errorOccured)
+
+}
+
+// createVisibilities creates visibilities from platform. Returns true if error has occurred
+func (r ReconciliationTask) createVisibilities(visibilities []*platform.ServiceVisibilityEntity) bool {
+	var errorOccured int32 // will be used as atomic bool
+	var wg sync.WaitGroup
+
+	for _, visibility := range visibilities {
+		execAsync(&wg, visibility, &errorOccured, r.createVisibility)
+	}
+	return await(&wg, &errorOccured)
+}
+
+func execAsync(wg *sync.WaitGroup, visibility *platform.ServiceVisibilityEntity, errorOccured *int32, f func(*platform.ServiceVisibilityEntity) error) {
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		if err := f(visibility); err != nil {
```

Comment:
any of these functions must be cancellable either via a quit channel, or via a context with cancellation (timeout)